internal/keepers/service.go

```diff
@@ -132,15 +143,90 @@ func (s *onDemandUpkeepService) CheckUpkeep(ctx context.Context, keys ...types.U
 func (s *onDemandUpkeepService) start() {
 	// TODO: if this process panics, restart it
 	go s.cacheCleaner.Run(s.cache)
+	go func() {
+		if err := s.runSamplingUpkeeps(); err != nil {
+			s.logger.Fatal(err)
+		}
+	}()
 }
 
 func (s *onDemandUpkeepService) stop() {
 	close(s.stopProcs)
 	close(s.cacheCleaner.stop)
 }
 
-func (s *onDemandUpkeepService) parallelCheck(ctx context.Context, keys []types.UpkeepKey, filters ...func(types.UpkeepKey) bool) ([]*types.UpkeepResult, error) {
-	samples := newSyncedArray[*types.UpkeepResult]()
+func (s *onDemandUpkeepService) runSamplingUpkeeps() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	heads := make(chan types.BlockKey, 1)
+	defer close(heads)
+
+	// Start the sampling upkeep process for heads
+	go func() {
+		for head := range heads {
+			s.handleIncomingHead(ctx, head)
+		}
+	}()
+
+	// Cancel context when receiving the stop signal
+	go func() {
+		<-s.stopProcs
+		cancel()
+	}()
+
+	return s.headSubscriber.OnNewHead(ctx, func(blockKey types.BlockKey) {
+		// This is needed in order to do not block the process when a new head comes in.
+		// The running upkeep sampling process should be finished first before starting
+		// sampling for the next head.
+		select {
+		case heads <- blockKey:
+		default:
+		}
+	})
+}
+
+// handleIncomingHead performs checking upkeep logic for all eligible keys of the given head
+func (s *onDemandUpkeepService) handleIncomingHead(ctx context.Context, head types.BlockKey) {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
```

Comment:
Have we considered an alternative where we cancel the context once we get a new head?

It has some implications on fast chains that we'll end up cancelling quickly but can try to mitigate that by having a 'minimum' time for processing head. The benefit I would see is that we will process upkeeps on fresh head as soon as we receive it.

cc @EasterTheBunny

Reply:
This is what we have right now. We don't cancel the running job on a new head.

Reply:
What would be the tradeoffs for both approaches? We should think through this

Reply:
also the 10 seconds seems arbitrary. Should think through it

Reply:
I don't think we should cancel when a new head comes in. It doesn't provide any value to start over for every head. We can just run again with the latest head even with some gaps. This is a lot like what we're going to be doing in observability due to the same RPC latency issues we see there.